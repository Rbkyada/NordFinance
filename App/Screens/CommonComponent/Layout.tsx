import React from 'react';
import {
  KeyboardAvoidingView,
  StatusBar,
  StyleSheet,
  View,
  SafeAreaView,
  RefreshControl,
  StyleProp,
  ViewStyle,
  TextStyle,
  ImageStyle,
  ScrollView,
} from 'react-native';
import { isIOS } from '@Utils/Constant';
import CommonStyle from '@Theme/CommonStyle';
import LightTheme from '@Theme/LightTheme';

interface LayoutProps {
  children: React.ReactNode;
  title?: string;
  titleCenter?: boolean;
  titleTextStyle?: StyleProp<TextStyle>;
  titleNumberOfLines?: number;
  titleMaxLength?: number;
  padding?: number;
  bounces?: boolean;
  submit?: {
    onSubmit?: () => void;
    isSubmitProcessing?: boolean;
    submitTitle?: string;
    submitBtnStyle?: StyleProp<ViewStyle>;
    onSubmitBtnType?: 'btn' | 'img' | 'text' | 'custom';
    customSubmitComponent?: JSX.Element;
    submitImage?: string;
    submitImageStyle?: StyleProp<ImageStyle>;
  };
  scrollable?: boolean;
  backgroundColor?: string;
  showBack?: boolean;
  refreshControl?: {
    refreshing: boolean;
    onRefresh: () => void;
  };
  navBarContainerStyle?: StyleProp<ViewStyle>;
  removeContainerView?: boolean;
}

const Layout = (props: LayoutProps) => {
  const {
    children,
    padding = 10,
    scrollable = false,
    refreshControl,
    removeContainerView = false,
    bounces = true,
  } = props;

  return (
    <SafeAreaView style={[CommonStyle.flex1]} testID="safeContainer">
      <StatusBar backgroundColor={LightTheme.themeColor} />
      <KeyboardAvoidingView
        behavior="padding"
        style={styles.keyboardView}
        keyboardVerticalOffset={isIOS ? 0 : -500}>
        {(scrollable && (
          <ScrollView
            showsVerticalScrollIndicator={false}
            keyboardShouldPersistTaps="always"
            bounces={bounces}
            contentContainerStyle={[styles.scrollContainer, { padding }]}
            testID="scrollContainer"
            refreshControl={
              (refreshControl && (
                <RefreshControl
                  refreshing={refreshControl.refreshing}
                  onRefresh={refreshControl.onRefresh}
                  tintColor={LightTheme.themeColor}
                />
              )) ||
              undefined
            }>
            {children}
          </ScrollView>
        )) ||
          (removeContainerView && (
            <View style={{ padding }}>{children}</View>
          )) || (
            <View style={[CommonStyle.flex1, { padding }]}>{children}</View>
          )}
      </KeyboardAvoidingView>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  keyboardView: {
    flex: 1,
    backgroundColor: 'transparent',
  },
  scrollContainer: { flexGrow: 1 },
});

export { Layout };
