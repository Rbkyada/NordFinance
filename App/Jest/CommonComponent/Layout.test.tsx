import renderer from 'react-test-renderer';
import { Layout } from '@CommonComponent/Layout';
import { Text, View } from 'react-native';

const props = {
  children: <></>,
};

const component = renderer.create(<Layout {...props} />);

describe('Layout component', () => {
  test('Layout component render', () => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('Layout render correctly', () => {
    const tree = component.toJSON();
    expect(tree).toBeTruthy();
  });

  test('safeContainer test', () => {
    const tree = component.root.findByProps({ testID: 'safeContainer' });
    expect(tree).toBeTruthy();
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly without scrollable', () => {
    const tree = renderer
      .create(
        <Layout>
          <View>
            <Text>Content</Text>
          </View>
        </Layout>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with scrollable', () => {
    const tree = renderer
      .create(
        <Layout scrollable={true}>
          <View>
            <Text>Content</Text>
          </View>
        </Layout>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with refresh control', () => {
    const refreshControl = {
      refreshing: false,
      onRefresh: jest.fn(),
    };
    const tree = renderer
      .create(
        <Layout scrollable={true} refreshControl={refreshControl}>
          <View>
            <Text>Content</Text>
          </View>
        </Layout>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
