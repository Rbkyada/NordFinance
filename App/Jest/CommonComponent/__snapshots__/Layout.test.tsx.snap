// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout component Layout component render 1`] = `
<RCTSafeAreaView
  style={
    [
      {
        "flex": 1,
      },
    ]
  }
  testID="safeContainer"
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "backgroundColor": "transparent",
          "flex": 1,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "flex": 1,
          },
          {
            "padding": 10,
          },
        ]
      }
    />
  </View>
</RCTSafeAreaView>
`;

exports[`Layout component renders correctly with refresh control 1`] = `
<RCTSafeAreaView
  style={
    [
      {
        "flex": 1,
      },
    ]
  }
  testID="safeContainer"
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "backgroundColor": "transparent",
          "flex": 1,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      bounces={true}
      contentContainerStyle={
        [
          {
            "flexGrow": 1,
          },
          {
            "padding": 10,
          },
        ]
      }
      keyboardShouldPersistTaps="always"
      refreshControl={
        <RefreshControlMock
          onRefresh={[MockFunction]}
          refreshing={false}
          tintColor="rgba(158, 172, 219, 1)"
        />
      }
      showsVerticalScrollIndicator={false}
      testID="scrollContainer"
    >
      <RCTRefreshControl />
      <View>
        <View>
          <Text>
            Content
          </Text>
        </View>
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;

exports[`Layout component renders correctly with scrollable 1`] = `
<RCTSafeAreaView
  style={
    [
      {
        "flex": 1,
      },
    ]
  }
  testID="safeContainer"
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "backgroundColor": "transparent",
          "flex": 1,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <RCTScrollView
      bounces={true}
      contentContainerStyle={
        [
          {
            "flexGrow": 1,
          },
          {
            "padding": 10,
          },
        ]
      }
      keyboardShouldPersistTaps="always"
      showsVerticalScrollIndicator={false}
      testID="scrollContainer"
    >
      <View>
        <View>
          <Text>
            Content
          </Text>
        </View>
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;

exports[`Layout component renders correctly without scrollable 1`] = `
<RCTSafeAreaView
  style={
    [
      {
        "flex": 1,
      },
    ]
  }
  testID="safeContainer"
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "backgroundColor": "transparent",
          "flex": 1,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "flex": 1,
          },
          {
            "padding": 10,
          },
        ]
      }
    >
      <View>
        <Text>
          Content
        </Text>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
